### PR TITLE
Leverage TE wgrad accumulation fusion in MFSDP v2

### LINKED ISSUE #6942: Leverage TE wgrad accumulation fusion in MFSDP v2
MFSDP v2 currently materializes local wgrads, then copies them into the reduce-scatter input buffer.

In `experimental/module.py::_reduce_gradient_groups()`, MFSDP v2 allocates the partial-gradient buffer and calls `copy_gradients_to_partial_buffer()`. `experimental/parameter_group.py::copy_gradients_to_partial_buffer()` then copies each `unsharded.grad` into the corresponding partial-buffer view before reduce-scatter.

Transformer Engine wgrad accumulation fusion can write wgrads directly into a preallocated destination buffer. Using the MFSDP v2 reduce-scatter input views as that destination should eliminate this extra full-gradient copy, as well as the separate transient autograd-gradient buffer.

TE enables this through its wgrad-accumulation option and a preallocated `main_grad`-style destination.

**Done when:** supported TE linear layers write directly to MFSDP v2’s reduce-scatter input buffer; gradients match the existing path; and a benchmark reports memory and step-time impact.

### PR DESCRIPTION
**Pull request by PACTHEMAN123**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Leverage Transformer Engine wgrad accumulation fusion in MFSDP v2 so supported TE/MCore linear layers write weight gradients directly into MFSDP v2 reduce-scatter input buffers.

The implementation follows the MFSDP runtime schedule described in `megatron/core/distributed/fsdp/src/docs/runtime_schedule.md`.

## Issue tracking

Linked issue: Fixes #6942

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code
- [ ] I have added relevant documentation
- [ ] I have run the autoformatter on my PR

The test implementations used for validation are intentionally kept outside this production-only PR. The validation results are reported below.

## Implementation

- Allow supported TE/MCore fused linear layers to write directly into preallocated MFSDP v2 reduce-scatter input buffers.
- Allocate and zero fused-wgrad destinations on the reduce-scatter stream.
- Preserve accumulation semantics when the same parameter is used by multiple fused backward paths, including MTP and the main GPT output layer.
- Pass explicit accumulation behavior through MCore and Transformer Engine wgrad GEMM paths.
- Avoid TE first-microbatch overwrite behavior for MFSDP-managed parameters.
- Keep ordinary autograd gradients and tied input/output contributions supported.
- Preserve backward all-gather prefetch ordering and stream/event dependencies from the MFSDP runtime schedule.

MFSDP v2 fused wgrad accumulation rejects unsupported combinations:

- Transformer Engine older than 2.10
- CPU offloading
- `overlap_moe_expert_parallel_comm`
- `overlap_dispatch_backward_with_experts_wgrad`
- `use_transformer_engine_op_fuser`
- NCCL-UB/symmetric-memory communication

The delayed expert-wgrad overlap combination is intentionally rejected because MFSDP v2 does not currently implement the delayed `post_wgrad_grad_acc_hook` callback required by `MoELayer.backward_dw()`.

## Validation

- Native 2-layer GPT with 2 MTP layers completed baseline and fused training runs.
- Main LM loss, both MTP losses, and gradient norms matched at printed precision.

## Benchmark

Configuration:

- 8x NVIDIA H20
- 8.03B-parameter GPT model
- 32 layers, hidden size 4096, FFN size 14336
- Sequence length 2048
- MFSDP v2 with pure data parallelism
- Two microbatches per rank
- Global batch size 16
- 15 iterations, with the first 5 excluded as warmup
- Reported step time is the trimmed mean of iterations 6-15

| Metric                       |      Baseline |         Fused |    Delta |
| ---------------------------- | ------------: | ------------: | -------: |
| Step time                    |   1674.150 ms |   1653.875 ms |  -1.211% |
| Aggregate throughput         | 19572.9 tok/s | 19812.9 tok/s |  +1.226% |
| Peak allocated memory        |   24518.26 MB |   24746.81 MB |  +0.932% |
| Peak reserved memory         |      30656 MB |      30072 MB |  -1.905% |
| Allocated after an iteration |   11586.54 MB |   11588.01 MB | +1.47 MB |

The first-step language-model loss was `12.58627` for both paths. Both runs completed all 15 iterations without OOM or NCCL errors, and all eight GPUs were released after completion.

## Limitations

MFSDP v2 continues to use the static backward-order prefetch design described in `runtime_schedule.md`.

Delayed expert wgrad overlap and CPU offloading remain unsupported when fused wgrad accumulation is enabled.

### CODE REVIEW

**Review by wujingyue — CHANGES_REQUESTED** [review]
Thanks for the PR! 

For the first PR, don't touch mcore_fsdp_adapter or things outside megatron_fsdp. 

Add unit tests. 

Simplify -- quite some slops at this moment.

**Comment by PACTHEMAN123** [comment]
> Thanks for the PR!
> 
> For the first PR, don't touch mcore_fsdp_adapter or things outside megatron_fsdp.
> 
> Add unit tests.
> 
> Simplify -- quite some slops at this moment.

Thanks, that makes sense. I’ll narrow this PR to the MFSDP-side buffer plumbing and focused unit tests. I’ll leave the adapter and MCore/TE integration, including the end-to-end correctness checks and benchmark, for a follow-up PR.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/parameter_group.py [hunk 1]
```diff
@@ -241,6 +262,20 @@ def __init__(
...
+                parameter.get_main_grad = _get_main_grad.__get__(parameter, type(parameter))
+                parameter.main_grad = None
+                # The destination is zeroed before backward, so fused kernels should
+                # accumulate into it.
+                parameter.overwrite_main_grad = False
+                # TE returns a shared zero dummy gradient so PyTorch runs the
+                # parameter-completion hook without materializing the real wgrad.
+                parameter.grad_added_to_main_grad = False
+                parameter.zero_out_wgrad = True
+                setattr(parameter, _PARAMETER_GROUP_INDEX_ATTR, index)
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/parameter_group.py:278]: I've been trying to avoid parameter attributes because they are fragile. 
1. They are not part of the nn.Parameter API. 
2. They increase the coupling between MFSDP and TE's **implementation**. 
3. User code occasionally replaces parameters via `module._parameters[key]`. It's not trivial to carry over these attributes from old parameters to new. 

Can you remind me what's the minimum to enable wgrad accumulation fusion (TE only; no MFSDP at all) so we can explore the options?

**Review by wujingyue — CHANGES_REQUESTED** [review]

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/parameter_group.py (hunk 1 above)
Comment (PACTHEMAN123) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/parameter_group.py:278]: For standalone TE, the minimum requirement is `fuse_wgrad_accumulation=True` on the module and a preallocated `weight.main_grad` before the wgrad GEMM runs:

```python
linear = te.Linear(..., fuse_wgrad_accumulation=True)
linear.weight.main_grad = torch.zeros_like(
    linear.weight,
    dtype=torch.float32,
)
```

With the default `is_first_microbatch=None`, TE accumulates into `main_grad`. `overwrite_main_grad=True` is optional and changes this to overwrite behavior.

For MFSDP, the remaining difficulty is providing that buffer without disrupting the runtime schedule. The reduce-scatter input should be allocated and released on the reduce-scatter stream, while the wgrad GEMM writes to it on the default stream. We can record an event after allocation and zeroing, but the default stream should wait for that event only when TE is about to launch the wgrad GEMM. Waiting in the module pre-backward hook would be too early and could serialize `Bwd_i` behind `RS_{i+1}`; not waiting would introduce a race.

I should have separated the minimal TE requirement from this MFSDP scheduling problem more clearly. Sorry for the confusion. A clean integration likely needs a TE-side destination provider or callback at the wgrad boundary, through which MFSDP can return the buffer view and establish the event dependency without adding custom state to `nn.Parameter`.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/parameter_group.py:278]: There seems to be a big gap between
```
linear = te.Linear(..., fuse_wgrad_accumulation=True)
linear.weight.main_grad = torch.zeros_like(
    linear.weight,
    dtype=torch.float32,
)
```
and the complicated logic in this PR. I've yet to be convinced this is the right solution.
